pkg/storaged/cockpit-clevis-tool.py

```diff
@@ -0,0 +1,162 @@
+#! /usr/bin/python3
+
+# This is a huge hack, let's not bother to make it nice and efficient.
+
+import sys
+import os
+import json
+import subprocess
+import base64
+import hashlib
+import re
+
+import urllib.request
+
+def b64_decode(data):
+    # The data we get doesn't seem to have any padding, but the base64
+    # module requires it.  So we add it back.  Can't anyone agree on
+    # anything?  Not even base64?
+    return base64.urlsafe_b64decode(data + '=' * ((4 - len(data) % 4) % 4))
+
+def decode_clevis_slot(dev, slot):
+    out = subprocess.check_output([ "luksmeta", "load", "-d", dev, "-s", str(slot) ])
+    data = json.loads(b64_decode(out.decode().split(".")[0]))
+    if data['clevis']['pin'] == "tang":
+        return {
+            "slot": slot,
+            "type": "tang",
+            "url": data['clevis']['tang']['url'],
+            "key": data['kid'],
+            "sigkeys": list(map(compute_thp, filter(is_signing_key, data['clevis']['tang']['adv']['keys']))),
+        }
+
+def info(dev):
+    result = subprocess.run([ "luksmeta", "show", "-d", dev ], stdout=subprocess.PIPE, stderr=subprocess.DEVNULL)
+    slots = [ ]
+    if result.returncode != 0:
+        return None
+    for line in result.stdout.splitlines():
+        fields = re.split(b" +", line)
+        if fields[1] == b"active" and fields[2] == b"cb6e8904-81ff-40da-a84a-07ab9ab5715e":
+            info = decode_clevis_slot(dev, int(fields[0]))
+            if info:
+                slots.append(info)
+    return slots
+
+def infos():
+    res = { }
+    for dev in subprocess.check_output([ "lsblk", "-pln", "-o", "NAME" ]).splitlines():
+        i = info(dev)
+        if i:
+            res[dev.decode()] = i
+    return res
+
+def monitor():
+    mon = subprocess.Popen([ "stdbuf", "-o", "L", "udevadm", "monitor", "-u", "-s", "block"],
+                           bufsize=1, stdout=subprocess.PIPE)
```

Comment:
Urgh, I hope that will go away -- udisks devices or systemd device units (both D-Bus accessible through signals) don't have the information you need? This doesn't even use `--env` to show the properties, so it looks like this literally just watches for new/removed block devices?

Reply:
Yes, this will go away.  This is just to get change notifications when luksmeta has written something new to a disk.  When UDisks2 does it, it will fit nicely into its existing udev event loop.